golem-worker-executor-base/src/services/worker_fork.rs

```diff
@@ -0,0 +1,472 @@
+            worker_proxy,
+            worker_enumeration_service,
+            running_worker_enumeration_service,
+            promise_service,
+            golem_config,
+            shard_service,
+            key_value_service,
+            blob_store_service,
+            oplog_service,
+            scheduler_service,
+            worker_activator,
+            events,
+            file_loader,
+            plugins,
+            oplog_processor_plugin,
+            extra_deps,
+        }
+    }
+
+    async fn validate_worker_forking(
+        &self,
+        account_id: &AccountId,
+        source_worker_id: &WorkerId,
+        target_worker_id: &WorkerId,
+        oplog_index_cut_off: OplogIndex,
+    ) -> Result<(OwnedWorkerId, OwnedWorkerId), GolemError> {
+        let second_index = OplogIndex::INITIAL.next();
+
+        if oplog_index_cut_off < second_index {
+            return Err(GolemError::invalid_request(
+                "oplog_index_cut_off must be at least 2",
+            ));
+        }
+
+        let owned_target_worker_id = OwnedWorkerId::new(account_id, target_worker_id);
+
+        let target_metadata = self.worker_service.get(&owned_target_worker_id).await;
+
+        // We allow forking only if the target worker does not exist
+        if target_metadata.is_some() {
+            return Err(GolemError::worker_already_exists(target_worker_id.clone()));
+        }
+
+        // We assume the source worker belongs to this executor
+        self.shard_service.check_worker(source_worker_id)?;
+
+        let owned_source_worker_id = OwnedWorkerId::new(account_id, source_worker_id);
+
+        self.worker_service
+            .get(&owned_source_worker_id)
+            .await
+            .ok_or(GolemError::worker_not_found(source_worker_id.clone()))?;
+
+        Ok((owned_source_worker_id, owned_target_worker_id))
+    }
+}
+
+#[async_trait]
+impl<Ctx: WorkerCtx> WorkerForkService for DefaultWorkerFork<Ctx> {
+    async fn fork(
+        &self,
+        source_worker_id: &OwnedWorkerId,
+        target_worker_id: &WorkerId,
+        oplog_index_cut_off: OplogIndex,
+    ) -> Result<(), GolemError> {
+        record_worker_call("fork");
+
+        let (owned_source_worker_id, owned_target_worker_id) = self
+            .validate_worker_forking(
+                &source_worker_id.account_id,
+                &source_worker_id.worker_id,
+                target_worker_id,
+                oplog_index_cut_off,
+            )
+            .await?;
+
+        let target_worker_id = owned_target_worker_id.worker_id.clone();
+        let account_id = owned_target_worker_id.account_id.clone();
+
+        let source_worker_instance =
+            Worker::get_or_create_suspended(self, &owned_source_worker_id, None, None, None, None)
+                .await?;
+
+        let source_worker_metadata = source_worker_instance.get_metadata().await?;
+
+        // Not sure if we should copy the metadata or not, or stick on to just default
```

Comment:
Let's remove this comment :)

Reply:
Added that comment myself, before understanding the entire code